File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        """
        the bottleneck of water trapped inside would be the minimum of the left and right max
        """
        l,r = 0,len(height)-1
        leftMax = height[l]
        rightMax = height[r]
        result = 0

        # process the entire elevation map
        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax,height[l])
                result += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax,height[r])
                result += rightMax - height[r]

        return result

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution 
{
public:
    int trap(vector<int>& height) 
    {
        /*
        *   to get how much water it can fill at a certain index, you subtract the min of leftMax and rightMax with the current height.
        *   min of leftMax and rightMax determines the bound since water can only fill <= min(leftMax,rightMax)
        *
        *   smaller boundary is the limiting factor because water would
        *   spill over that side first.
        */
        int l {}, r {static_cast<int>(height.size()-1)};
        int leftMax {height[l]}, rightMax {height[r]};
        int result {};

        while(l<r)
        { 
            // if left max is the smaller bound, shift to right and update result
            if(leftMax < rightMax)
            {
                l++;
                leftMax = max(leftMax, height[l]);   // update leftMax 
                result += leftMax - height[l];
            }
            else  // if right max is the smaller bound, shift to left and update result
            {
                r--;
                rightMax = max(rightMax, height[r]); // update rightMax 
                result += rightMax - height[r];
            }
        }

        return result;
    }
};